### PR TITLE
feat: Auditing group members as part of group resource

### LINKED ISSUE #4736: audit: updating users in a group should generate an audit log diff
When one updates a group by adding or removing users, we show an 'update' audit log, but there is no diff generated:
Screen Shot 2022-10-24 at 6 06 57 PM

We should show the users added or removed.

### PR DESCRIPTION
**Pull request by Kira-Pilot**
resolves #4736 

We now audit group members as part of group changes:
Screen Shot 2023-01-17 at 2 58 07 PM

### CODE REVIEW

File: enterprise/audit/diff.go [hunk 1]
```diff
@@ -27,6 +27,8 @@ func diffValues(left, right any, table Table) audit.Map {
 		diffKey = table[structName(rightT)]
 	)
 
+	fmt.Println("DIFF KEY", diffKey)
+
```
Comment (Kira-Pilot) [inline enterprise/audit/diff.go:31]: This is logging as the `AuditableGroup` map which I think is right.

File: enterprise/audit/diff.go [hunk 2]
```diff
@@ -45,6 +47,9 @@ func diffValues(left, right any, table Table) audit.Map {
 
 			diffName = rightT.Field(i).Tag.Get("json")
 		)
+		fmt.Println("rightT.Field(i)", rightT, rightT.Field(i), rightT.Field(i).Tag.Get("json"))
+
+		fmt.Println("DIFF NAME", diffName)
```
Comment (Kira-Pilot) [inline enterprise/audit/diff.go:52]: This is undefined

File: enterprise/audit/diff.go [hunk 3]
```diff
@@ -45,6 +47,9 @@ func diffValues(left, right any, table Table) audit.Map {
 
 			diffName = rightT.Field(i).Tag.Get("json")
 		)
+		fmt.Println("rightT.Field(i)", rightT, rightT.Field(i), rightT.Field(i).Tag.Get("json"))
```
Comment (Kira-Pilot) [inline enterprise/audit/diff.go:50]: I don't think this json tag is logging.

File: enterprise/audit/table.go [hunk 4]
```diff
@@ -128,6 +128,14 @@ var AuditableResources = auditMap(map[any]map[string]Action{
 		"reason":              ActionIgnore,
 		"daily_cost":          ActionIgnore,
 	},
+	&database.AuditableGroup{}: {
+		"id":              ActionTrack,
+		"name":            ActionTrack,
+		"organization_id": ActionIgnore, // Never changes.
+		"avatar_url":      ActionTrack,
+		"quota_allowance": ActionTrack,
+		"members":         ActionTrack,
```
Comment (Kira-Pilot) [inline enterprise/audit/table.go:137]: AuditableGroup now contains members

File: codersdk/audit.go [hunk 5]
```diff
@@ -22,7 +22,13 @@ const (
...
 	ResourceTypeGitSSHKey       ResourceType = "git_ssh_key"
 	ResourceTypeAPIKey          ResourceType = "api_key"
-	ResourceTypeGroup           ResourceType = "group"
+	// ResourceTypeGroup           ResourceType = "group"
+
+	// ResourceTypeAuditableGroup ResourceType = "auditable_group" // does not work
+	// ResourceTypeAuditableGroup ResourceType = "group" // does not work
+
+	ResourceTypeGroup ResourceType = "auditable_group"
+	// ResourceTypeGroup	ResourceType = "group"
```
Comment (Kira-Pilot) [inline codersdk/audit.go:31]: Something about this is off I think

File: coderd/database/modelmethods.go [hunk 6]
```diff
@@ -1,9 +1,36 @@
...
 import (
+	"sort"
+
 	"github.com/coder/coder/coderd/rbac"
 )
 
+type AuditableGroup struct {
+	Group
+	Members []GroupMember
+}
```
Comment (kylecarbs) [inline coderd/database/modelmethods.go:12]: Everything auditable needs to have JSON struct tags on it.

```
Members []GroupMember `json:"members"
```

File: coderd/database/modelmethods.go [hunk 7]
```diff
@@ -1,9 +1,36 @@
...
+
 	"github.com/coder/coder/coderd/rbac"
 )
 
+type AuditableGroup struct {
+	Group
+	Members []GroupMember `json:"members"`
+}
+
+func (g Group) Auditable(users []User) AuditableGroup {
```
Comment (Emyrk) [inline coderd/database/modelmethods.go:14]: Comment this function

```suggestion
// Auditable returns an object that can be used in audit logs.
// Covers both group and group member changes.
func (g Group) Auditable(users []User) AuditableGroup {
```

File: coderd/database/modelmethods.go [hunk 8]
```diff
@@ -1,9 +1,36 @@
...
+	members := make([]GroupMember, 0, len(users))
+	for _, u := range users {
+		members = append(members, GroupMember{
+			UserID:  u.ID,
+			GroupID: g.ID,
+		})
+	}
+
+	// we sort to ensure the diff order enterprise/audit/diff.go:18
```
Comment (Emyrk) [inline coderd/database/modelmethods.go:23]:
```suggestion
	// consistent ordering
```

File: enterprise/coderd/groups.go [hunk 9]
```diff
@@ -257,15 +266,16 @@ func (api *API) deleteGroup(rw http.ResponseWriter, r *http.Request) {
...
 			Audit:   *auditor,
 			Log:     api.Logger,
 			Request: r,
 			Action:  database.AuditActionDelete,
 		})
 	)
 	defer commitAudit()
-	aReq.Old = group
+	var emptyUsers []database.User
+	aReq.Old = group.Auditable(emptyUsers)
```
Comment (Emyrk) [inline enterprise/coderd/groups.go:278]: shouldn't the `aReq.New` be `emptyUsers` and the `aReq.Old` have the members?

Comment (Kira-Pilot) [reply enterprise/coderd/groups.go]: Great catch @Emyrk! Updated.

File: enterprise/coderd/groups.go [hunk 10]
```diff
@@ -233,15 +242,15 @@ func (api *API) patchGroup(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	members, err := api.Database.GetGroupMembers(ctx, group.ID)
-	if err != nil {
+	patchedMembers, patchedMembersErr := api.Database.GetGroupMembers(ctx, group.ID)
+	if patchedMembersErr != nil {
```
Comment (coadler) [inline enterprise/coderd/groups.go:246]: Do these errors need to be named differently? It seems like the wrong error is being passed below.

**Review by coadler — APPROVED** [review]

File: enterprise/coderd/groups.go (hunk 10 above)
Comment (Kira-Pilot) [reply enterprise/coderd/groups.go:246]: Oops! Will fix. What convention do you like for naming errors? Just call them `err` until there's a scoping conflict?

Comment (coadler) [reply enterprise/coderd/groups.go:246]: Unless you need multiple errors in scope at once, I would always just name them `err`!

Comment (Kira-Pilot) [reply enterprise/coderd/groups.go:246]: Sounds good! Will update now.

File: site/src/components/AuditLogRow/AuditLogRow.tsx [hunk 11]
```diff
@@ -49,6 +54,25 @@ export const AuditLogRow: React.FC<AuditLogRowProps> = ({
     ? `${browser.name} ${browser.version}`
     : t("auditLog:table.logRow.notAvailable")
 
+  let auditDiff = auditLog.diff
+  // groups have nested diffs (group members)
+  // so we overwrite the member diff such that
+  // only the user_id is shown.
+  if (auditLog.resource_type === "group") {
```
Comment (presleyp) [inline site/src/components/AuditLogRow/AuditLogRow.tsx:61]: Could you pull this out into a util function, maybe with a unit test?

**Review by presleyp — APPROVED** [review]

**Pull request merged**